File: python/internal.py
import params
from adrs import ADRS, toInt
from fors import fors_sign, fors_pkFromSig
from hypertree import ht_sign, ht_verify
from shake import H_msg, PRF_msg
from xmss import xmss_node


# algorithm 18
def slh_keygen_internal(sk_seed: bytes, sk_prf: bytes, pk_seed: bytes) -> tuple:
    adrs = ADRS()
    adrs.setLayerAddress(params.prm.d - 1)
    pk_root = xmss_node(sk_seed, 0, params.prm.h_, pk_seed, adrs)
    return ((sk_seed + sk_prf + pk_seed + pk_root), (pk_seed + pk_root))


# algorithm 19
def slh_sign_internal(M: bytes, SK: bytes, addrnd: bytes) -> bytes:
    # precompute these values to make the code cleaner
    index1 = (params.prm.k * params.prm.a + 7) // 8
    index2 = ((params.prm.h - (params.prm.h // params.prm.d)) + 7) // 8
    index3 = ((params.prm.h // params.prm.d) + 7) // 8

    sk_seed = SK[0               :params.prm.n]
    sk_prf  = SK[    params.prm.n:2 * params.prm.n]
    pk_seed = SK[2 * params.prm.n:3 * params.prm.n]
    pk_root = SK[3 * params.prm.n:]

    SIG = b""
    adrs = ADRS()

    R = PRF_msg(sk_prf, addrnd, M)
    SIG += R
    digest = H_msg(R, pk_seed, pk_root, M)
    md = digest[0:index1]
    tmp_idx_tree = digest[index1:index1 + index2]
    tmp_idx_leaf = digest[index1 + index2:index1 + index2 + index3]

    idx_tree = toInt(tmp_idx_tree, index2) % (2 ** (params.prm.h - (params.prm.h // params.prm.d)))
    idx_leaf = toInt(tmp_idx_leaf, index3) % (2 ** (params.prm.h // params.prm.d))

    adrs.setTreeAddress(idx_tree)
    adrs.setTypeAndClear(params.prm.FORS_TREE)
    adrs.setKeyPairAddress(idx_leaf)

    SIG_FORS = fors_sign(md, sk_seed, pk_seed, adrs)
    SIG += SIG_FORS

    PK_FORS = fors_pkFromSig(SIG_FORS, md, pk_seed, adrs)
    SIG_HT = ht_sign(PK_FORS, sk_seed, pk_seed, idx_tree, idx_leaf)
    SIG += SIG_HT
    return SIG


# algorithm 20
def slh_verify_internal(M: bytes, SIG: bytes, PK: bytes) -> bool:
    index1 = (params.prm.k * params.prm.a + 7) // 8
    index2 = ((params.prm.h - (params.prm.h // params.prm.d)) + 7) // 8
    index3 = ((params.prm.h // params.prm.d) + 7) // 8

    pk_seed = PK[0:params.prm.n]
    pk_root = PK[params.prm.n:]

    if len(SIG) != (1 + params.prm.k * (1 + params.prm.a) + params.prm.h + params.prm.d * params.prm.len) * params.prm.n:
        print("Invalid signature length")
        return False

    adrs = ADRS()
    R = SIG[0:params.prm.n]
    SIG_FORS = SIG[params.prm.n:(1 + params.prm.k * (1 + params.prm.a)) * params.prm.n]
    SIG_HT = SIG[(1 + params.prm.k * (1 + params.prm.a)) * params.prm.n:]

    digest = H_msg(R, pk_seed, pk_root, M)
    md = digest[0:index1]
    tmp_idx_tree = digest[index1:index1 + index2]
    tmp_idx_leaf = digest[index1 + index2:index1 + index2 + index3]
    idx_tree = toInt(tmp_idx_tree, index2) % (2 ** (params.prm.h - (params.prm.h // params.prm.d)))
    idx_leaf = toInt(tmp_idx_leaf, index3) % (2 ** (params.prm.h // params.prm.d))

    adrs.setTreeAddress(idx_tree)
    adrs.setTypeAndClear(params.prm.FORS_TREE)
    adrs.setKeyPairAddress(idx_leaf)

    PK_FORS = fors_pkFromSig(SIG_FORS, md, pk_seed, adrs)
    return ht_verify(PK_FORS, SIG_HT, pk_seed, idx_tree, idx_leaf, pk_root)

File: C/internal.c
#include <stdbool.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include "adrs.h"
#include "fors.h"
#include "hypertree.h"
#include "params.h"
#include "shake.h"
#include "xmss.h"

// algorithm 18
void slh_keygen_internal(Parameters *prm, uint8_t *sk_seed, uint8_t *sk_prf, uint8_t *pk_seed, uint8_t *SK, uint8_t *PK)
{
    ADRS adrs;
    initADRS(&adrs);

    setLayerAddress(&adrs, prm->d - 1);
    uint8_t pk_root[prm->n];
    xmss_node(prm, sk_seed, 0, prm->h_, pk_seed, adrs, pk_root);

    memcpy(SK + 0 * prm->n, sk_seed, prm->n);
    memcpy(SK + 1 * prm->n, sk_prf,  prm->n);
    memcpy(SK + 2 * prm->n, pk_seed, prm->n);
    memcpy(SK + 3 * prm->n, pk_root, prm->n);

    memcpy(PK + 0 * prm->n, pk_seed, prm->n);
    memcpy(PK + 1 * prm->n, pk_root, prm->n);
}

// algorithm 19
void slh_sign_internal(Parameters *prm, uint8_t *M, size_t M_len, const uint8_t *SK, const uint8_t *addrnd, uint8_t *buffer)
{
    // precompute these values to make the code cleaner
    uint64_t index1 = (prm->k * prm->a + 7) / 8;
    uint64_t index2 = ((prm->h - (prm->h / prm->d)) + 7) / 8;
    uint64_t index3 = ((prm->h / prm->d) + 7) / 8;

    ADRS adrs;
    initADRS(&adrs);

    uint8_t sk_seed[prm->n];
    uint8_t sk_prf[prm->n];
    uint8_t pk_seed[prm->n];
    uint8_t pk_root[prm->n];
    memcpy(sk_seed, SK + 0 * prm->n, prm->n);
    memcpy(sk_prf,  SK + 1 * prm->n, prm->n);
    memcpy(pk_seed, SK + 2 * prm->n, prm->n);
    memcpy(pk_root, SK + 3 * prm->n, prm->n);

    uint32_t sig_fors_len = prm->k * (1 + prm->a) * prm->n;
    uint32_t sig_ht_len = (prm->h + prm->d * prm->len) * prm->n;
    // signature = Randomness + FORS signature + HT signature
    uint8_t SIG[prm->n + sig_fors_len + sig_ht_len];

    // Generate R using PRF
    uint8_t R[prm->n];
    PRF_msg(prm, sk_prf, addrnd, M, M_len, R);
    memcpy(SIG, R, prm->n);

    // Generate message digest
    uint8_t digest[prm->m];
    H_msg(prm, R, pk_seed, pk_root, M, M_len, digest);

    uint8_t md[index1];
    memcpy(md, digest, index1);

    uint8_t tmp_idx_tree[index2];
    uint8_t tmp_idx_leaf[index3];
    memcpy(tmp_idx_tree, digest + index1, index2);
    memcpy(tmp_idx_leaf, digest + index1 + index2, index3);

    uint64_t idx_tree = toInt(tmp_idx_tree, index2) & (UINT64_C(1) << (prm->h - prm->h_)) - UINT64_C(1);
    uint64_t idx_leaf = toInt(tmp_idx_leaf, index3) & (1 << prm->h_) - 1;

    setTreeAddress(&adrs, idx_tree);
    setTypeAndClear(&adrs, prm->FORS_TREE);
    setKeyPairAddress(&adrs, idx_leaf);

    // Generate FORS signature
    uint8_t SIG_FORS[sig_fors_len];
    fors_sign(prm, md, sk_seed, pk_seed, adrs, SIG_FORS);

    // Copy FORS signature to main signature
    memcpy(SIG + prm->n, SIG_FORS, sig_fors_len);

    uint8_t PK_FORS[prm->n];
    fors_pkFromSig(prm, SIG_FORS, md, pk_seed, adrs, PK_FORS);

    // Generate and append HT signature
    uint8_t SIG_HT[sig_ht_len];
    ht_sign(prm, PK_FORS, sk_seed, pk_seed, idx_tree, idx_leaf, SIG_HT);
    memcpy(SIG + prm->n + sig_fors_len, SIG_HT, sig_ht_len);
    memcpy(buffer, SIG, prm->n + sig_fors_len + sig_ht_len);
}

// algorithm 20
bool slh_verify_internal(Parameters *prm, uint8_t *M, size_t M_len, uint8_t *SIG, size_t SIG_len, const uint8_t *PK)
{
    uint64_t index1 = (prm->k * prm->a + 7) / 8;
    uint64_t index2 = ((prm->h - (prm->h / prm->d)) + 7) / 8;
    uint64_t index3 = ((prm->h / prm->d) + 7) / 8;

    uint32_t sig_fors_len = prm->k * (1 + prm->a) * prm->n;
    uint32_t sig_ht_len = (prm->h + prm->d * prm->len) * prm->n;
    uint32_t sig_len = prm->n + sig_fors_len + sig_ht_len;

    uint8_t pk_seed[prm->n];
    uint8_t pk_root[prm->n];
    memcpy(pk_seed, PK + 0 * prm->n, prm->n);
    memcpy(pk_root, PK + 1 * prm->n, prm->n);

    if (SIG_len != sig_len) {
        printf("Signature has invalid length\n");
        return false;
    }

    ADRS adrs;
    initADRS(&adrs);

    uint8_t R[prm->n];
    memcpy(R, SIG, prm->n);

    // Extract FORS and HT signatures
    uint8_t SIG_FORS[sig_fors_len];
    memcpy(SIG_FORS, SIG + prm->n, sig_fors_len);

    uint8_t SIG_HT[sig_ht_len];
    memcpy(SIG_HT, SIG + prm->n + sig_fors_len, sig_ht_len);

    uint8_t digest[prm->m];
    H_msg(prm, R, pk_seed, pk_root, M, M_len, digest);
    uint8_t md[index1];
    memcpy(md, digest, index1);

    uint8_t tmp_idx_tree[index2];
    uint8_t tmp_idx_leaf[index3];
    memcpy(tmp_idx_tree, digest + index1, index2);
    memcpy(tmp_idx_leaf, digest + index1 + index2, index3);

    uint64_t idx_tree = toInt(tmp_idx_tree, index2) & (UINT64_MAX >> (64 - (prm->h - prm->h / prm->d)));
    uint64_t idx_leaf = toInt(tmp_idx_leaf, index3) & (UINT64_MAX >> (64 - prm->h / prm->d));

    setTreeAddress(&adrs, idx_tree);
    setTypeAndClear(&adrs, prm->FORS_TREE);
    setKeyPairAddress(&adrs, idx_leaf);

    uint8_t PK_FORS[prm->n];
    fors_pkFromSig(prm, SIG_FORS, md, pk_seed, adrs, PK_FORS);

    return ht_verify(prm, PK_FORS, SIG_HT, pk_seed, idx_tree, idx_leaf, pk_root);
}
